shards/egui/src/visual/mod.rs

```diff
@@ -0,0 +1,2521 @@
+    }
+  }
+
+  fn get_mut(&mut self) -> Option<&mut mpsc::Receiver<T>> {
+    self.receiver.as_mut()
+  }
+}
+
+impl_custom_any!(FunctionState);
+
+#[derive(Debug, Clone, PartialEq)]
+struct SequenceState {
+  selected: bool,
+}
+
+impl_custom_any!(SequenceState);
+
+// common state
+pub struct Context {
+  swap_state: Option<SwapState>,
+  seqs_zoom_stack: Vec<*mut Sequence>,
+  seqs_stack: Vec<*mut Sequence>,
+  has_changed: bool,
+}
+
+pub struct VisualAst<'a> {
+  context: &'a mut Context,
+  ui: &'a mut Ui,
+  parent_selected: bool,
+}
+
+// Helper function to determine if a color is light
+fn is_light_color(color: egui::Color32) -> bool {
+  let brightness = 0.299 * color.r() as f32 + 0.587 * color.g() as f32 + 0.114 * color.b() as f32;
+  brightness > 128.0
+}
+
+fn emoji(s: &str) -> egui::RichText {
+  egui::RichText::new(s)
+    .family(egui::FontFamily::Proportional)
+    // .strong()
+    .size(14.0)
+}
+
+fn get_first_shard_ref<'a>(ast: &'a mut Sequence) -> Option<&'a mut Function> {
+  for statement in &mut ast.statements {
+    match statement {
+      Statement::Assignment(assignment) => {
+        if let Assignment::AssignRef(pipeline, _) = assignment {
+          if let BlockContent::Shard(shard) = &mut pipeline.blocks[0].content {
+            return Some(shard);
+          }
+        }
+      }
+      Statement::Pipeline(pipeline) => {
+        if let BlockContent::Shard(shard) = &mut pipeline.blocks[0].content {
+          return Some(shard);
+        }
+      }
+    }
+  }
+  None
+}
+fn get_last_shard_ref<'a>(ast: &'a mut Sequence) -> Option<&'a mut Function> {
+  if ast.statements.len() < 2 {
+    return None;
+  }
+  for statement in ast.statements.iter_mut().rev() {
+    match statement {
+      Statement::Assignment(assignment) => {
+        if let Assignment::AssignRef(pipeline, _) = assignment {
+          if let BlockContent::Shard(shard) = &mut pipeline.blocks.last_mut()?.content {
+            return Some(shard);
+          }
+        }
+      }
+      Statement::Pipeline(pipeline) => {
+        if let BlockContent::Shard(shard) = &mut pipeline.blocks.last_mut()?.content {
+          return Some(shard);
+        }
+      }
+    }
+  }
+  None
+}
+
+impl<'a> VisualAst<'a> {
+  pub fn new(context: &'a mut Context, ui: &'a mut Ui) -> Self {
+    VisualAst {
+      context,
+      ui,
+      parent_selected: false,
+    }
+  }
+
+  pub fn with_parent_selected(
+    context: &'a mut Context,
+    ui: &'a mut Ui,
+    parent_selected: bool,
+  ) -> Self {
+    VisualAst {
+      context,
+      ui,
+      parent_selected,
+    }
+  }
+
+  pub fn render(&mut self, ast: &mut Sequence) -> Option<Response> {
+    ast.accept_mut(self)
+  }
+
+  fn mutate_shard(&mut self, x: &mut Function) -> Option<Response> {
+    let state = x.get_or_insert_custom_state(|| FunctionState {
+      params_sorted: false,
+    });
+    let params_sorted = state.params_sorted;
+
+    // // check if we have a result from a pending operation
```

Comment:
Remove comment

Reply:
Was keeping those as memo on how to use it once I had to actually